src/react/reconcilation.js

```diff
@@ -277,41 +285,163 @@ export class Reconciler {
     return classMetadata;
   }
 
-  _renderRelayQueryRendererComponent(
+  _resolveContextProviderComponent(
     componentType: Value,
     reactElement: ObjectValue,
-    props: ObjectValue | AbstractValue | AbstractObjectValue,
     context: ObjectValue | AbstractObjectValue,
+    branchStatus: BranchStatusEnum,
+    branchState: BranchState | null,
     evaluatedNode: ReactEvaluatedNode
-  ) {
-    let renderResult = {
-      result: reactElement,
-      childContext: context,
+  ): Value {
+    let typeValue = Get(this.realm, reactElement, "type");
+    let propsValue = Get(this.realm, reactElement, "props");
+
+    let evaluatedChildNode = createReactEvaluatedNode("NORMAL", "Context.Provider");
+    evaluatedNode.children.push(evaluatedChildNode);
+    this.statistics.componentsEvaluated++;
+    let newBranchState = new BranchState();
+    invariant(typeValue instanceof ObjectValue || typeValue instanceof AbstractObjectValue);
+    const contextConsumer = Get(this.realm, typeValue, "context");
+    invariant(contextConsumer instanceof ObjectValue || contextConsumer instanceof AbstractObjectValue);
+    let lastValueProp = getProperty(this.realm, contextConsumer, "currentValue");
+
+    const setContextCurrentValue = value => {
+      if (value instanceof Value) {
+        // update the currentValue
+        setProperty(this.realm, contextConsumer, "currentValue", value);
+      }
     };
+    // if we have a value prop, set it
+    if (propsValue instanceof ObjectValue || propsValue instanceof AbstractObjectValue) {
+      let valueProp = Get(this.realm, propsValue, "value");
+      setContextCurrentValue(valueProp);
+    }
+
+    newBranchState.addContext(contextConsumer);
+    if (this.componentTreeConfig.firstRenderOnly) {
```

Comment:
Why is this first render only? If we know the value (which might be abstract), it should be fine for updates too. Since it's abstract it'll be able to update.

Reply:
I wanted to be on the safe side right now for updates, as we don't know if a parent in an unknown parent tree exists.